tests/async/test_browsertype_connect.py

```diff
@@ -0,0 +1,142 @@
+    await page.goto(server.EMPTY_PAGE)
+    await browser.close()
+
+
+async def test_browser_type_connect_should_be_able_to_connect_two_browsers_at_the_same_time(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser1 = await browser_type.connect(remote_server.ws_endpoint)
+    assert len(browser1.contexts) == 0
+    await browser1.new_context()
+    assert len(browser1.contexts) == 1
+
+    browser2 = await browser_type.connect(remote_server.ws_endpoint)
+    assert len(browser2.contexts) == 0
+    await browser2.new_context()
+    assert len(browser2.contexts) == 1
+    assert len(browser1.contexts) == 1
+
+    await browser1.close()
+    page2 = await browser2.new_page()
+    # original browser should still work
+    assert await page2.evaluate("7 * 6") == 42
+
+    await browser2.close()
+
+
+async def test_browser_type_connect_disconnected_event_should_be_emitted_when_browser_is_closed_or_server_is_closed(
+    browser_type: BrowserType, launch_server
+):
+    # Launch another server to not affect other tests.
+    remote = launch_server()
+
+    browser1 = await browser_type.connect(remote.ws_endpoint)
+    browser2 = await browser_type.connect(remote.ws_endpoint)
+
+    disconnected1 = []
+    disconnected2_event_future: asyncio.Future[bool] = asyncio.Future()
+    browser1.on("disconnected", lambda: disconnected1.append(True))
+    browser2.on("disconnected", lambda: disconnected2_event_future.set_result(True))
+
+    page2 = await browser2.new_page()
+
+    await browser1.close()
+    assert len(disconnected1) == 1
+    assert disconnected2_event_future.done() is False
+
+    remote.kill()
+    assert len(disconnected1) == 1
+
+    with pytest.raises(Error):
+        # Tickle connection so that it gets a chance to dispatch disconnect event.
+        await page2.title()
+    await disconnected2_event_future
+
+
+async def test_browser_type_disconnected_event_should_have_browser_as_argument(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = await browser_type.connect(remote_server.ws_endpoint)
+    event_payloads = []
+    browser.on("disconnected", lambda b: event_payloads.append(b))
+    await browser.close()
+    assert event_payloads[0] == browser
+
+
+async def test_browser_type_connect_set_browser_connected_state(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = await browser_type.connect(remote_server.ws_endpoint)
+    assert browser.is_connected()
+    await browser.close()
+    assert browser.is_connected() is False
+
+
+async def test_browser_type_connect_should_throw_when_used_after_is_connected_returns_false(
+    browser_type: BrowserType, launch_server
+):
+    remote_server = launch_server()
+    browser = await browser_type.connect(remote_server.ws_endpoint)
+    page = await browser.new_page()
+
+    remote_server.kill()
+
+    with pytest.raises(Error) as exc_info:
```

Comment:
Consider using `match` with `raises` https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions

Reply:
Interesting, but going to keep it for now for consistency reasons in the codebase.

Reply:
:+1: It can be changed later in a atomic change in the codebase